### DISCUSSION TITLE
Cross-reference scoping and qualified names

### DISCUSSION THREAD

**Discussion by JoanGi** [eclipse-langium/langium#336] [Q&A] [answered]
Hi!,

First, thanks for this great initiative, it will be great to see how evolves this tool and I will provide as much as feed-back I could in my learning process. The case is that I'm trying  to build a DSL to define requirements over data with this tool, and I've not previous experience with Xtext, so maybe I missing some point.

**The problem:** 
Its about scoping and cross-reference: I can make a cross-reference (with autocompletion) if the references entity is at the same level, but not if it is inside another one. I know that I have to reference to the fully qualified name (at leat, I've read this in Xtext issues), but I'm not able to figure out how to accomplish it with Langium. 

In the code above, at the "Requeriments" declaration, the cross-referece reporter[Author] works, but  desname=[Instance|QualifiedName] don't.  The qualified name of the reference will be [The Dataset ID].[The Instance ID], d'ont?

Also, I want to reference attributes, which I really need [The Dataset NAME].[The Instance NAME].[The Attribute NAME]

```// Definitions
Model:
    (declarations+=Declaration)*;

Declaration:
    (Author |DatasetDefinition | Requeriment);

// Authors
Author:
    'author' name=ID 'email:' email=ID;

// Datasets
DatasetDefinition:
    'dataset' name=ID 'withDescription:' description=ID 'composedBy:' (instances+=Instance)*;

Instance: 
    'instance' name=ID 'ofType:'  type=Type 'withAttributes:' (attributes+=Attribute)*;

Attribute:
    'attribute' name=ID 'ofType' attType=ATtype ;

// Requeriments
Requeriment:
    'requeriment' name=ID 'reportedby:' reporter=[Author] 'fromAtt' desName=[Instance|QualifiedName] ;

// Data Types Rules
QualifiedName returns string: ID ('.' ID)*;
ATtype returns string: 'int' | 'string' | 'bool';
Type returns string: 'people' | 'images' | 'raw';
```

Thanks!! and another time, congratulation for the work!!

pd: I've attach a example of my DSL. :)

![image](https://user-images.githubusercontent.com/5485515/148088396-b6a42f63-133e-4242-b979-bcd3f5c38c4c.png)

**Comment by msujew** [comment, marked as answer]
Hi @JoanGi, thank you for your question and the compliments :)

One major difference to our scoping implementation compared to Xtext is that we left the default implementation mostly empty. Like you noticed it works in a local scope, but there's no fully qualified name referencing mechanism. We opted for this mostly for two reasons, 1. its easier ;) 2. with Xtext the most popular use case was building programming languages, which use fully qualified names only in selected scenarios, usually relying on the local and global scope otherwise.

Anyway, you can take a look at the [domainmodel example scoping](https://github.com/langium/langium/blob/main/examples/domainmodel/src/language-server/domain-model-scope.ts), if you want to achieve fully qualified name scoping behavior.

**Comment by JoanGi** [comment]
Thanks for the quick answer!. 

Finally I've get it working with my grammar. I've faced some problems to solve, so I want to share my thought, maybe are helpful.

There is no fully qualified names referencing mechanism, but the default implementation also do not reference elements out of the scope by its name. In [domain model example index](https://github.com/langium/langium/blob/main/examples/domainmodel/src/language-server/domain-model-index.ts) the example fix it using the function StreamAllContent() instead of StreamContent().

I've been looking at the [DefaultAstNodeDescriptionProvider ](https://github.com/langium/langium/blob/main/packages/langium/src/workspace/ast-descriptions.ts ) and I've tried to change it, solving this issue. I've been trying for unwanted behaviors, cross-file reference still works. 

In my case, it was a need, maybe for other cases this could behave in an unexpected manner. Maybe developing a way to express in the grammar if the scope of the reference is local or global could be a solution. 

Thanks!!
